server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationState.java

```diff
@@ -0,0 +1,485 @@
+    }
+
+    public VotingConfiguration getLastAcceptedConfiguration() {
+        return getLastAcceptedState().getLastAcceptedConfiguration();
+    }
+
+    public long getLastPublishedVersion() {
+        return lastPublishedVersion;
+    }
+
+    public boolean electionWon() {
+        return electionWon;
+    }
+
+    public boolean isElectionQuorum(VoteCollection votes) {
+        return votes.isQuorum(getLastCommittedConfiguration()) && votes.isQuorum(getLastAcceptedConfiguration());
+    }
+
+    public boolean isPublishQuorum(VoteCollection votes) {
+        return votes.isQuorum(getLastCommittedConfiguration()) && votes.isQuorum(lastPublishedConfiguration);
+    }
+
+    public boolean containsJoinVoteFor(DiscoveryNode node) {
+        return joinVotes.containsVoteFor(node);
+    }
+
+    public boolean joinVotesHaveQuorumFor(VotingConfiguration votingConfiguration) {
+        return joinVotes.isQuorum(votingConfiguration);
+    }
+
+    /**
+     * Used to bootstrap a cluster by injecting the initial state and configuration.
+     *
+     * @param initialState The initial state to use. Must have term 0, version 1, and non-empty configurations.
+     * @throws CoordinationStateRejectedException if the arguments were incompatible with the current state of this object.
+     */
+    public void setInitialState(ClusterState initialState) {
+        final long lastAcceptedVersion = getLastAcceptedVersion();
+        if (lastAcceptedVersion != 0) {
+            logger.debug("setInitialState: rejecting since last-accepted version {} > 0", lastAcceptedVersion);
+            throw new CoordinationStateRejectedException("initial state already set: last-accepted version now " + lastAcceptedVersion);
+        }
+
+        assert getLastAcceptedTerm() == 0;
+        assert getLastAcceptedConfiguration().isEmpty();
+        assert getLastCommittedConfiguration().isEmpty();
+        assert lastPublishedVersion == 0;
+        assert lastPublishedConfiguration.isEmpty();
+        assert electionWon == false;
+        assert joinVotes.isEmpty();
+        assert publishVotes.isEmpty();
+
+        assert initialState.term() == 0;
+        assert initialState.version() == 1;
+        assert initialState.getLastAcceptedConfiguration().isEmpty() == false;
+        assert initialState.getLastCommittedConfiguration().isEmpty() == false;
+
+        persistedState.setLastAcceptedState(initialState);
+    }
+
+    /**
+     * May be safely called at any time to move this instance to a new term.
+     *
+     * @param startJoinRequest The startJoinRequest, specifying the node requesting the join.
+     * @return A Join that should be sent to the target node of the join.
+     * @throws CoordinationStateRejectedException if the arguments were incompatible with the current state of this object.
+     */
+    public Join handleStartJoin(StartJoinRequest startJoinRequest) {
+        if (startJoinRequest.getTerm() <= getCurrentTerm()) {
```

Comment:
Tests all pass if this is `<` instead of `<=`.

Reply:
Did you run the tests multiple times? Both testStartJoinBeforeBootstrap and testStartJoinAfterBootstrap fail approx. 30% of the times if this is changed from `<=` to `<`.

Reply:
Only a couple of times, not enough clearly.

Reply:
I've pushed 3bec606f179 🗡